### PR TITLE
8273494: Zero: Put libjvm.so into "zero" folder, not "server"

### PR DESCRIPTION
**Pull request by shipilev**
Currently, the build system defaults the libjvm.so location to "server".  This makes looking for `libjvm.so` awkward, see JDK-8273487 for example. We need to see if moving the libjvm.so to a proper location breaks anything. 

Additional testing:
 - [x] Linux x86_64 Zero build
 - [x] Linux x86_64 Zero bootcycle-images
 - [x] Linux x86_64 Zero `tier1`

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8273494](https://bugs.openjdk.java.net/browse/JDK-8273494): Zero: Put libjvm.so into "zero" folder, not "server"

### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**)
 * [Severin Gehwolf](https://openjdk.java.net/census#sgehwolf) (@jerboaa - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5440/head:pull/5440` \
`$ git checkout pull/5440`

Update a local copy of the PR: \
`$ git checkout pull/5440` \
`$ git pull https://git.openjdk.java.net/jdk pull/5440/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5440`

View PR using the GUI difftool: \
`$ git pr show -t 5440`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5440.diff">https://git.openjdk.java.net/jdk/pull/5440.diff</a>

</details>

### CODE REVIEW

File: make/autoconf/hotspot.m4
Comment (dholmes-ora) [inline make/autoconf/hotspot.m4:86]: I presume "zero" was a special variant? Are there any other variants remaining?

**Review by dholmes-ora — COMMENTED** [review]
This does what it intends but I can't approve it via a review as the issue is more about whether this should be done. The people who own/maintain/support Zero are the ones who should be directing this.

David

File: make/autoconf/hotspot.m4
Comment (shipilev) [reply make/autoconf/hotspot.m4:86]: Yes, there are at least "core" and "custom":

```
# All valid JVM variants
VALID_JVM_VARIANTS="server client minimal core zero custom"
```

Comment (magicus) [reply make/autoconf/hotspot.m4:86]: I think we should stop these as well from impersonating the server JVM. Preferably in the same fix, so we can remove all the special casing for "server" being anything else but server.

**Review by magicus — COMMENTED** [review]
And as for @dholmes-ora comment: I'm not sure who "own" zero at this point in time. Aleksey has made a lot of the zero patches lately; does that not count? Are you thinking about any specific person that needs to weight in on this?

File: make/autoconf/hotspot.m4
Comment (shipilev) [reply make/autoconf/hotspot.m4:86]: Ok, I agree. Can I do a Zero-specific thing here (so that it is potentially cleanly backportable), and then handle the rest of the variants?

Comment (magicus) [reply make/autoconf/hotspot.m4:86]: Sure, that works for me.

Comment (shipilev) [reply make/autoconf/hotspot.m4:86]: While we are at it, do `core` and `custom` even carry their weight? I cannot remember if I ever seen anyone using them. Maybe we should "just" drop those variants, and leave only "server, client, minimal, zero"?

Comment (magicus) [reply make/autoconf/hotspot.m4:86]: I don't know about "core". Last time it was up for discussion, some old-timer I've forgotten pointed out that it was useful for such-and-such. I'm not sure if the cost for keeping it is high enough to make the effort to figure out if anyone is going to miss it if we remove it.

As for "custom", it is not a real JVM configuration per se. Instead, it's more of a testing ground for the build system. Basically, the JVM "variants" are just named pre-defined sets of enabled JVM features. To test this properly, I introduced the "custom" variant which is just an empty set of JVM features from the start, so that individual features can be fully controlled. I don't really think anyone is using this in the "real world", but I'd like to keep it for testing purposes.

**Comment by dholmes-ora** [comment]
@magicus every "port" in OpenJDK is supposed to have a clear owner and support system. Zero has been somewhat lacking in that area but there were enough people to keep it surviving. Now I'm not so sure. Does Zero have an active user community? Developer community? If so they are the ones who need to assess this change.

**Comment by shipilev** [comment]
> @magicus every "port" in OpenJDK is supposed to have a clear owner and support system. Zero has been somewhat lacking in that area but there were enough people to keep it surviving. Now I'm not so sure. Does Zero have an active user community? Developer community? If so they are the ones who need to assess this change.

FWIW, Zero had fallen into Red Hat hands for support. The official lead (if you look at Census) is Gary Benson, who is not active in this project anymore. Since then, it was mostly supported by RH folks (like me) with contributions from Debian, Tencent, Huawei, Alibaba -- mostly because all of us have arch ports that do not have full-blown Server VMs yet.

I have put things in motion to claim the leadership more formally. I do have to note, though, that over the last few years of me whipping Zero into shape, this is the first time anyone asked the formal governance question, which must tell us something about how much we care about ownership formalities here ;)

**Comment by dholmes-ora** [comment]
It isn't the "formal governance" I'm concerned about, more about the folk who use/rely on Zero being the ones to evaluate the impact of a change like this. People, like myself, who do not use Zero in any way cannot evaluate whether this change is appropriate - it's that simple. :)

**Comment by gnu-andrew** [comment]
> > @magicus every "port" in OpenJDK is supposed to have a clear owner and support system. Zero has been somewhat lacking in that area but there were enough people to keep it surviving. Now I'm not so sure. Does Zero have an active user community? Developer community? If so they are the ones who need to assess this change.
> 
> FWIW, Zero had fallen into Red Hat hands for support. The official lead (if you look at Census) is Gary Benson, who is not active in this project anymore. Since then, it was mostly supported by RH folks (like me) with contributions from Debian, Tencent, Huawei, Alibaba -- mostly because all of us have arch ports that do not have full-blown Server VMs yet.

For clarity, Gary was a part of the Red Hat Java team at the time, so it has always been a Red Hat project. At the time, the only JIT ports in OpenJDK were x86, x86-64 and SPARC, so the other architectures Red Hat supported (ppc, ppc64, s390, s390x) needed some way to at least build and run to allow OpenJDK packages to be shipped, even if end users desiring greater performance used some other JDK.

As OpenJDK has gained further JIT ports, the use of Zero has declined in response. For example, we can build OpenJDK 17 on every architecture we need without using Zero, so it's not going to receive that kind of "you broke our build" testing any more. It is still used for 8u & 11u, and is definitely worth keeping alive to help bootstrap any new ports.

So, in short, the owner has always been and remains Red Hat, even if the individual personnel have changed. The formal project represents a time when it was maintained outside the mainline OpenJDK, so there has been no need to update that for a long time, as any JDK project committer can make changes to the code in the main JDK project.

> 
> I have put things in motion to claim the leadership more formally. I do have to note, though, that over the last few years of me whipping Zero into shape, this is the first time anyone asked the formal governance question, which must tell us something about how much we care about ownership formalities here ;)

Most of the OpenJDK committers at Red Hat have pushed Zero fixes at some time, myself included, as need has arisen. This has the first time this has been raised to my knowledge.

Regarding this change itself, I think it's fine for trunk, where there's time to shake out any issues for OpenJDK 18, but I wouldn't want to backport it. These kind of changes tend to throw up things only when someone's application breaks, and, at least if that's a new JDK version, that's somewhat expected.

**Review by magicus — APPROVED** [review]
I'm changing my response to "approve". I agree that we can change zero only in this patch to facilitate backporting, and fix the remaining odds and ends in a separate PR.

**Comment by magicus** [comment]
And also let me be clear that I see no reason to delay pushing this any further. Everyone who has any vested interest in zero has had ample chance to comment on the PR.

**Review by jerboaa — APPROVED** [review]
Looks fine to me.

**Comment by shipilev** [comment]
I'll be happy to integrate as long as @dholmes-ora has no objections.

**Comment by dholmes-ora** [comment]
@shipilev no objections from me. I will leave it to the Zero developers to worry, or not, about the Zero users.

Cheers,
David

**Comment by shipilev** [comment]
Thank you all!

/integrate

**Pull request closed without merging**
